### PR TITLE
rewrite functions using new api

### LINKED ISSUE #225: 1.1.2 does not work with directory environments on Puppetserver 1.1.3 (due to puppet_x)
### Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 3.8.7 (Puppetserver 1.1.3-1.el7)
- Ruby: jruby (with puppetserver)
- Distribution: Centos 7 
- Module version: 1.1.2

### How to reproduce (e.g Puppet code you use)

Use environment name "myenvironment".

Puppet run failes with: 

```
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Could not autoload puppet/type/archive: Could not autoload puppet/provider/archive/curl: Could not find parent provider ruby of curl on node NODE
```

In this case puppet/archive is consumed by rtyler/jenkins 1.7.0 (https://github.com/jenkinsci/puppet-jenkins/releases/tag/v1.7.0).

### What are you seeing

Class loading failes, caused by https://tickets.puppetlabs.com/browse/SERVER-973 on Puppetserver. 

### What behaviour did you expect instead

Should compile.

### LINKED ISSUE #320: Could not autoload puppet/parser/functions/artifactory_sha1: no such file to load -- puppet_x/bodeco/util
Thats somehow annoying .. I cant progress with my work at the moment .. 

Whats wrong here ?

`
[root@qa5 disk2]# puppet agent -t --server puppet --environment cm
Info: Using configured environment 'cm'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Could not autoload puppet/parser/functions/artifactory_sha1: no such file to load -- puppet_x/bodeco/util at /etc/puppetlabs/code/modules/archive/manifests/artifactory.pp:103:14 at /etc/puppetlabs/code/modules/our_application/manifests/deployer.pp:53 on node qa5
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
`

[root@qa5 disk2]# puppet agent --version
5.3.3

### LINKED ISSUE #321: Convert functions to puppet 4 API
Converting the puppet 3 API functions to use the Puppet 4 API, should fix environment isolation issues people occasionally have.

For example, it will probably fix https://github.com/voxpupuli/puppet-archive/issues/320

### PR DESCRIPTION
**Pull request by applewiskey**
Rewrite functions using new function API
    Fixes #321 
    Fixes #320 
    Fixes #225 (?)

### CODE REVIEW

**Comment by bastelfreak** [comment]
Hi @applewiskey, thanks for the PR. Can you take a look at the failing tests on travis?

File: lib/puppet/functions/archive/go_md5.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Public: go file md5 checksum
+#
+# args[0] - username
+# args[1] - password
+# args[2] - file_name
+# args[3] - go md5 checksum url
```
Comment (alexjfisher) [inline lib/puppet/functions/archive/go_md5.rb]: Could you convert this documentation to the new puppet strings format?  eg like https://github.com/voxpupuli/puppet-archive/blob/ab996a12193497c13c3a2e151915652289cd1e0b/lib/puppet/functions/archive/artifactory_checksum.rb#L5-L9

**Comment by alexjfisher** [comment]
@applewiskey Thanks for taking this on!  The tests will need moving/updating.  This shouldn't be too tricky.  Take a look at https://github.com/voxpupuli/puppet-archive/pull/323/files and ping me if you get stuck and need help.

Will the manifests that call these functions will also need updating to use the new namespace?

File: lib/puppet/functions/archive/go_md5.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+# Returns specific file's md5 from go server md5 checksum file
+
+Puppet::Functions.create_function(:'archive::go_md5') do
+  require 'puppet_x/bodeco/util'
+
+  dispatch :main do
+    required_param 'String', :username
+    required_param 'String', :password
+    required_param 'String', :file
+    required_param 'String', :url
```
Comment (ekohl) [inline lib/puppet/functions/archive/go_md5.rb]: Should this use `Stdlib::Httpurl` or are protocols other than http(s) also supported?

File: lib/puppet/functions/archive/assemble_nexus_url.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Puppet::Functions.create_function(:'archive::assemble_nexus_url') do
+  require "cgi"
+
+  dispatch :assemble_nexus_url do
+    required_param 'String', :nexus_url
+    required_param 'Hash', :params
+  end
+
+  def assemble_nexus_url(nexus_url, params)
+    service_relative_url = 'service/local/artifact/maven/content'
```
Comment (ekohl) [inline lib/puppet/functions/archive/assemble_nexus_url.rb]: Should this be a param with a default too?

**Comment by applewiskey** [comment]
Sorry for a delay.
I don't get why first 3 tests failing and what should I come up with reading this
```
      Failure/Error: Puppet.features.stubs(:root?).returns(true)
      
      Mocha::NotInitializedError:
        Mocha methods cannot be used outside the context of a test
```

Anyway, it doesn't fix #320

File: lib/puppet/functions/archive/assemble_nexus_url.rb (hunk 3 above)
Comment (applewiskey) [reply lib/puppet/functions/archive/assemble_nexus_url.rb]: I don't it should.

File: lib/puppet/functions/archive/go_md5.rb (hunk 2 above)
Comment (applewiskey) [reply lib/puppet/functions/archive/go_md5.rb]: It checked against String in define archive::nexus anyway

**Comment by bastelfreak** [comment]
@applewiskey the issues should go away after a rebase. They are caused by a specific version of the mocha gem.

**Comment by ekohl** [comment]
Could you rebase and fix the rubocop error?

**Comment by bastelfreak** [comment]
I get the following errors locally after rebasing:

```
Failures:

  1) go_md5 retreives file md5
     Failure/Error: expect(scope.function_go_md5(['user', 'pass', 'filea', url])).to eq '283158c7da8c0ada74502794fa8745eb'
     
     NoMethodError:
       undefined method `function_go_md5' for Scope():Puppet::Parser::Scope
     # ./.vendor/ruby/2.5.0/gems/puppet-5.5.3/lib/puppet/parser/scope.rb:1037:in `method_missing'
     # ./spec/unit/puppet/parser/functions/go_md5_spec.rb:16:in `block (2 levels) in <top (required)>'
```

**Comment by ekohl** [comment]
Also adding the backwards-incompatible flag because this removes the old functions and users could rely on that. I'm leaning to keeping them in for now so it's not a major version bump.

**Comment by bastelfreak** [comment]
ping @applewiskey :)

**Comment by applewiskey** [comment]
Rebased. Anything else?

**Comment by bastelfreak** [comment]
@applewiskey please take a look at the failing travis jobs.

**Comment by applewiskey** [comment]
I fixed rubocop, but I don't really know how to fix something like this
```
F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 1)
F    should contain File[/opt/app/example.zip] with owner => "0", group => "0", mode => "0640" and require => "Archive[/opt/app/example.zip]" (FAILED - 2)
  go archive with path
F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 3)
F    should contain File[/opt/app/example.zip] with owner => "app", group => "app", mode => "0400" and require => "Archive[/opt/app/example.zip]" (FAILED - 4)
```
and a lot more different errors. Errors not in code I changed.

**Comment by alexjfisher** [comment]
> I fixed rubocop, but I don't really know how to fix something like this
> 
> ```
> F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 1)
> F    should contain File[/opt/app/example.zip] with owner => "0", group => "0", mode => "0640" and require => "Archive[/opt/app/example.zip]" (FAILED - 2)
>   go archive with path
> F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 3)
> F    should contain File[/opt/app/example.zip] with owner => "app", group => "app", mode => "0400" and require => "Archive[/opt/app/example.zip]" (FAILED - 4)
> ```
> 
> and a lot more different errors. Errors not in code I changed.

The spec test for the defined type was mocking the old non-namespaced function.  This needed to be [updated](https://github.com/voxpupuli/puppet-archive/pull/392/files#diff-a6dbe892af312a9f108f66d707514aa9L6-L10).

@applewiskey Thanks for your work on this, and sorry we didn't follow it through and get it merged at the time.

I've superceded this PR with 3 smaller ones.  I think I've covered everything you had, as well as getting the tests converted and passing.
https://github.com/voxpupuli/puppet-archive/pull/390
https://github.com/voxpupuli/puppet-archive/pull/391
https://github.com/voxpupuli/puppet-archive/pull/392

**Comment by applewiskey** [comment]
Closing as there are
#390
#391
#392
by @alexjfisher

**Pull request closed without merging**